### PR TITLE
8324585: JVM native memory leak in PCKS11-NSS security provider

### PR DESCRIPTION
**Pull request by djelinski**
Please review this patch that fixes a memory leak in P11TlsPrfGenerator, which is triggered during TLS1.2 Finished message generation and verification.

The patch changes C_SignInit JNI method to free the mechanism data immediately after use. This matches the behavior of other Init methods (like C_EncryptInit). The patch also fixes a similar issue in other signature-related methods.

The change essentially reverts part of [JDK-8080462](https://bugs.openjdk.org/browse/JDK-8080462).

All sun/security/pkcs11 tests still pass with NSS ~3.35 and~ 3.91. All tier1-3 tests still pass.

EDIT:
Some sun/security/pkcs11 tests fail with NSS 3.64 and older, see [comment](https://github.com/openjdk/jdk/pull/17584#issuecomment-1914665234)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324585](https://bugs.openjdk.org/browse/JDK-8324585): JVM native memory leak in PCKS11-NSS security provider (**Bug** - P4)

### Reviewers
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**) ⚠️ Review applies to [87edba10](https://git.openjdk.org/jdk/pull/17584/files/87edba103808da6bfc0de50a2eae87c48b894c54)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17584/head:pull/17584` \
`$ git checkout pull/17584`

Update a local copy of the PR: \
`$ git checkout pull/17584` \
`$ git pull https://git.openjdk.org/jdk.git pull/17584/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17584`

View PR using the GUI difftool: \
`$ git pr show -t 17584`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17584.diff">https://git.openjdk.org/jdk/pull/17584.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17584#issuecomment-1911793890)

### CODE REVIEW

**Comment by valeriepeng** [comment]
IIRC, this may be the special handling to work around the PSS errors I observed when implementing the support. Good that we don't need them now.

**Review by valeriepeng — APPROVED** [review]

**Comment by djelinski** [comment]
Updated the code to disable RSA-PSS on known bad NSS versions

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java [hunk 1]
```diff
@@ -1288,6 +1306,13 @@ private void initToken(CK_SLOT_INFO slotInfo) throws PKCS11Exception {
                 continue;
             }
 
+            if (brokenMechanisms.contains(longMech)) {
+                if (showInfo) {
+                    System.out.println("DISABLED due to broken provider");
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1311]: nit: how about changing "broken provider" to "known issue with NSS"?

**Review by valeriepeng — APPROVED** [review]
Just a minor wording suggestion. 
Disabling older NSS releases due to known issue is the best that we can do to work with multiple NSS releases out there.

**Comment by djelinski** [comment]
I applied your suggestion. Thanks for the review!

/integrate

**Pull request closed without merging**